diff --git a/node_modules/jstack-shared/dist/index.cjs b/node_modules/jstack-shared/dist/index.cjs
index 27d9524..2a14f0f 100644
--- a/node_modules/jstack-shared/dist/index.cjs
+++ b/node_modules/jstack-shared/dist/index.cjs
@@ -4352,9 +4352,11 @@ var EventEmitter = class {
   handleEvent(eventName, data) {
     const handlers = this.eventHandlers.get(eventName);
     if (!handlers?.length) {
-      logger.warn(
-        `No handlers registered for event "${eventName}". Did you forget to call .on("${eventName}", handler)?`
-      );
+      if (eventName !== "ping") {
+        logger.warn(
+          `No handlers registered for event "${eventName}". Did you forget to call .on("${eventName}", handler)?`
+        );
+      }
       return;
     }
     let validatedData = data;
@@ -4415,7 +4417,7 @@ var EventEmitter = class {
   on(event, callback) {
     if (!callback) {
       logger.error(
-        `No callback provided for event handler "${event.toString()}". Ppass a callback to handle this event.`
+        `No callback provided for event handler "${event.toString()}". Pass a callback to handle this event.`
       );
       return;
     }
diff --git a/node_modules/jstack-shared/dist/index.js b/node_modules/jstack-shared/dist/index.js
index 032fb2b..bfa0717 100644
--- a/node_modules/jstack-shared/dist/index.js
+++ b/node_modules/jstack-shared/dist/index.js
@@ -4343,9 +4343,11 @@ var EventEmitter = class {
   handleEvent(eventName, data) {
     const handlers = this.eventHandlers.get(eventName);
     if (!handlers?.length) {
-      logger.warn(
-        `No handlers registered for event "${eventName}". Did you forget to call .on("${eventName}", handler)?`
-      );
+      if (eventName !== "ping") {
+        logger.warn(
+          `No handlers registered for event "${eventName}". Did you forget to call .on("${eventName}", handler)?`
+        );
+      }
       return;
     }
     let validatedData = data;
@@ -4406,7 +4408,7 @@ var EventEmitter = class {
   on(event, callback) {
     if (!callback) {
       logger.error(
-        `No callback provided for event handler "${event.toString()}". Ppass a callback to handle this event.`
+        `No callback provided for event handler "${event.toString()}". Pass a callback to handle this event.`
       );
       return;
     }
